File: javascript/anagram/anagram.js
function Matcher(anagram) {
    this.anagram = anagram;
}

Matcher.prototype.isAnagram = function(word) {
    return this.anagram.toLowerCase() !== word.toLowerCase() 
        && this.anagram.toLowerCase().split('').sort().join() === word.toLowerCase().split('').sort().join();
}

Matcher.prototype.matches = function() {
    var words = normalizeArgs(arguments);
    var that = this;
    return words.filter(function(word) { return that.isAnagram(word); });
}

function normalizeArgs(args) {
    if (args[0] instanceof Array) {
        return args[0];
    } else {
        return Array.prototype.slice.call(args);
    }
}

module.exports = function(anagram) {
    return new Matcher(anagram);
}

File: scala/anagram/src/main/scala/Anagram.scala
/**
 * anagram matcher
 * @param word - initial word to match with possible anagrams
 */
class Anagram(val word: String) {

  def matches(anagrams: Seq[String]): Seq[String] = {
    anagrams.filter(anagram => {
      anagram.toLowerCase != word.toLowerCase &&
        anagram.toLowerCase.toCharArray.sorted.mkString == word.toLowerCase.toCharArray.sorted.mkString
    })
  }
}
